linux/LinuxProcessList.c

```diff
@@ -1003,14 +1003,14 @@ static int handleNetlinkMsg(struct nl_msg* nlmsg, void* linuxProcess) {
       memcpy(&stats, nla_data(nla_next(nla_data(nlattr), &rem)), sizeof(stats));
       assert(lp->super.pid == (pid_t)stats.ac_pid);
 
-      unsigned long long int timeDelta = stats.ac_etime * 1000 - lp->delay_read_time;
-      #define BOUNDS(x) (isnan(x) ? 0.0 : ((x) > 100) ? 100.0 : (x))
-      #define DELTAPERC(x,y) BOUNDS((float) ((x) - (y)) / timeDelta * 100)
-      lp->cpu_delay_percent = DELTAPERC(stats.cpu_delay_total, lp->cpu_delay_total);
-      lp->blkio_delay_percent = DELTAPERC(stats.blkio_delay_total, lp->blkio_delay_total);
-      lp->swapin_delay_percent = DELTAPERC(stats.swapin_delay_total, lp->swapin_delay_total);
-      #undef DELTAPERC
-      #undef BOUNDS
+      if (stats.ac_etime * 1000 > lp->delay_read_time) {
+         unsigned long long int timeDelta = stats.ac_etime * 1000 - lp->delay_read_time;
+         #define DELTAPERC(x, y) MINIMUM((float)saturatingSub(x, y) / timeDelta * 100.0f, 100.0f)
+         lp->cpu_delay_percent = DELTAPERC(stats.cpu_delay_total, lp->cpu_delay_total);
+         lp->blkio_delay_percent = DELTAPERC(stats.blkio_delay_total, lp->blkio_delay_total);
+         lp->swapin_delay_percent = DELTAPERC(stats.swapin_delay_total, lp->swapin_delay_total);
+         #undef DELTAPERC
```

Comment:
There's a change of behavior in my version of the code. When `timeDelta` is zero, the old code would set all `*_delay_percent` members to zero (due to the result of the `BOUNDS` macro). In my version, the percentage values would be left untouched. Should I go back to the old behavior of always setting the value?

Reply:
If changing the behaviour, I think the only correct change would be to use NAN there. The old behaviour would be fine too IMHO.

Reply:
I'm considering revising this patch. There are a few issues with the current one.

1. [The Linux Kernel documentation about struct taskstats](https://docs.kernel.org/accounting/taskstats-struct.html) stated that `cpu_delay_total`, `blkio_delay_total` and `swapin_delay_total` may wrap around on overflow. Although the wraparound is unlikely until 2^64 nanoseconds (≈584.5 years, much later than Y2038) of uptime, it is enough to say that using `saturatingSub` would be too defensive on our part. So I might just leave these values wrap around during subtraction as the kernel originally expected.
2. There is still a need to check if the "delta time" is zero, but I don't consider the case being an error. There is (theoretically) a possibility that we scan the process states too fast before the kernel updates the data and thus we get the exact same data as the last scan. My original idea was to keep the percentages computed in the last scan (not turn them to zeros or NaNs) is this case happens.
3. I wish the behaviors of handling "zero delta times" be consistent across platforms, so may tweak the platform codes of Solaris, DragonFlyBSD and others.

Reply:
Can you catalog the behaviours on the various platforms before the change? So we can compare and chose from that?

Reply:
Splitting this part to issue #1274 in order to simplify this patch of the PR.